File: src/core/classification/blast-radius.ts
/**
 * Blast-radius derivation for unresolved edges (query-time).
 *
 * Third orthogonal signal axis alongside:
 *   - category (extraction failure mode)
 *   - classification (semantic meaning of the gap)
 *   - blast_radius (how architecturally dangerous is this unresolved)
 *
 * Computed on demand — NOT persisted. The heuristic is derived from:
 *   - basis_code (determines receiver origin)
 *   - category (scoping: only defined for CALLS-family)
 *   - source node visibility (enclosing scope significance)
 *
 * Scope (first slice):
 *   - defined for CALLS-family unresolved edges only
 *   - other categories return { blastRadius: "not_applicable" }
 *   - entrypoint-path detection is deferred (uses visibility only)
 *
 * Naming: `local_like` is the honest catch-all for receivers
 * that are NOT import-bound, NOT same-file-declared, NOT runtime
 * globals, NOT `this`-bound. They are likely local variables,
 * parameters, destructured locals, closure captures, or
 * intermediate expression roots. The label does NOT overclaim
 * exact provenance.
 */

import { UnresolvedEdgeBasisCode } from "../diagnostics/unresolved-edge-classification.js";
import { UnresolvedEdgeCategory } from "../diagnostics/unresolved-edge-categories.js";

// ── Vocabulary ──────────────────────────────────────────────────────

export type ReceiverOrigin =
	| "this_bound"
	| "import_bound_external"
	| "import_bound_internal"
	| "same_file_declared"
	| "runtime_global"
	| "runtime_module"
	| "local_like"
	| "not_applicable";

export type EnclosingScopeSignificance =
	| "exported_public"
	| "internal_private"
	| "unknown_scope";

export type BlastRadiusLevel =
	| "low"
	| "medium"
	| "high"
	| "not_applicable";

export interface BlastRadiusAssessment {
	receiverOrigin: ReceiverOrigin;
	enclosingScopeSignificance: EnclosingScopeSignificance;
	blastRadius: BlastRadiusLevel;
}

// ── CALLS-family categories ─────────────────────────────────────────

const CALLS_FAMILY = new Set<string>([
	UnresolvedEdgeCategory.CALLS_FUNCTION_AMBIGUOUS_OR_MISSING,
	UnresolvedEdgeCategory.CALLS_OBJ_METHOD_NEEDS_TYPE_INFO,
	UnresolvedEdgeCategory.CALLS_THIS_METHOD_NEEDS_CLASS_CONTEXT,
	UnresolvedEdgeCategory.CALLS_THIS_WILDCARD_METHOD_NEEDS_TYPE_INFO,
]);

// ── Derivation ──────────────────────────────────────────────────────

/**
 * Derive blast-radius assessment for a single unresolved edge.
 *
 * Pure function. All inputs are plain data from the sample row.
 * Returns `not_applicable` for non-CALLS categories.
 */
export function deriveBlastRadius(input: {
	category: string;
	basisCode: string;
	sourceNodeVisibility: string | null;
}): BlastRadiusAssessment {
	if (!CALLS_FAMILY.has(input.category)) {
		return {
			receiverOrigin: "not_applicable",
			enclosingScopeSignificance: "unknown_scope",
			blastRadius: "not_applicable",
		};
	}

	const origin = deriveReceiverOrigin(input.basisCode);
	const scope = deriveEnclosingScopeSignificance(input.sourceNodeVisibility);
	const radius = computeRadius(origin, scope);

	return {
		receiverOrigin: origin,
		enclosingScopeSignificance: scope,
		blastRadius: radius,
	};
}

/**
 * Map basis_code to receiver origin.
 *
 * The basis_code already encodes what rule matched (or didn't).
 * This function makes that implicit knowledge explicit.
 */
function deriveReceiverOrigin(basisCode: string): ReceiverOrigin {
	switch (basisCode) {
		case UnresolvedEdgeBasisCode.THIS_RECEIVER_IMPLIES_INTERNAL:
			return "this_bound";
		case UnresolvedEdgeBasisCode.RECEIVER_MATCHES_EXTERNAL_IMPORT:
		case UnresolvedEdgeBasisCode.CALLEE_MATCHES_EXTERNAL_IMPORT:
			return "import_bound_external";
		case UnresolvedEdgeBasisCode.RECEIVER_MATCHES_INTERNAL_IMPORT:
		case UnresolvedEdgeBasisCode.CALLEE_MATCHES_INTERNAL_IMPORT:
		case UnresolvedEdgeBasisCode.SPECIFIER_MATCHES_PROJECT_ALIAS:
			return "import_bound_internal";
		case UnresolvedEdgeBasisCode.RECEIVER_MATCHES_SAME_FILE_SYMBOL:
		case UnresolvedEdgeBasisCode.CALLEE_MATCHES_SAME_FILE_SYMBOL:
			return "same_file_declared";
		case UnresolvedEdgeBasisCode.RECEIVER_MATCHES_RUNTIME_GLOBAL:
		case UnresolvedEdgeBasisCode.CALLEE_MATCHES_RUNTIME_GLOBAL:
			return "runtime_global";
		case UnresolvedEdgeBasisCode.SPECIFIER_MATCHES_RUNTIME_MODULE:
			return "runtime_module";
		case UnresolvedEdgeBasisCode.NO_SUPPORTING_SIGNAL:
			return "local_like";
		default:
			return "local_like";
	}
}

function deriveEnclosingScopeSignificance(
	visibility: string | null,
): EnclosingScopeSignificance {
	// Visibility values are lowercase in the DB (see model/types.ts Visibility).
	// "export" and "public" = externally reachable.
	// "private", "protected", "internal", null = not externally reachable.
	if (visibility === "export" || visibility === "public") return "exported_public";
	if (
		visibility === "private" ||
		visibility === "protected" ||
		visibility === "internal" ||
		visibility === null
	) {
		return "internal_private";
	}
	return "unknown_scope";
}

/**
 * Combine receiver origin + enclosing scope into a blast-radius level.
 *
 * Principles:
 *   - local_like + internal_private = LOW (function-local noise)
 *   - local_like + exported_public = MEDIUM (local uncertainty on a public path)
 *   - import_bound = MEDIUM (cross-module reference)
 *   - runtime_global / runtime_module = LOW (well-understood external)
 *   - same_file_declared = LOW (contained within file)
 *   - this_bound + internal = LOW, this_bound + exported = MEDIUM
 *
 * Entrypoint-path detection (→ HIGH override) is deferred to a
 * future slice. Currently no edge reaches HIGH.
 */
function computeRadius(
	origin: ReceiverOrigin,
	scope: EnclosingScopeSignificance,
): BlastRadiusLevel {
	switch (origin) {
		case "local_like":
			return scope === "exported_public" ? "medium" : "low";
		case "this_bound":
			return scope === "exported_public" ? "medium" : "low";
		case "import_bound_internal":
			return "medium";
		case "import_bound_external":
			return "low";
		case "same_file_declared":
			return "low";
		case "runtime_global":
		case "runtime_module":
			return "low";
		case "not_applicable":
			return "not_applicable";
	}
}

File: rust/crates/classification/src/blast_radius.rs
//! Blast-radius derivation for unresolved edges.
//!
//! Mirror of `src/core/classification/blast-radius.ts`. Pure
//! function, no I/O, no state. Computes a `BlastRadiusAssessment`
//! from three input signals: the edge's category, the classifier's
//! basis code, and the source node's visibility.
//!
//! Scope (first slice):
//!   - defined for CALLS-family unresolved edges only
//!   - other categories return `{ blastRadius: NotApplicable }`
//!   - entrypoint-path detection is deferred (uses visibility only)

use crate::types::{
	BlastRadiusAssessment, BlastRadiusLevel, EnclosingScopeSignificance,
	ReceiverOrigin, UnresolvedEdgeBasisCode, UnresolvedEdgeCategory,
};

/// Derive the blast-radius assessment for a single unresolved edge.
///
/// Mirror of `deriveBlastRadius` from `blast-radius.ts:76`.
///
/// Pure function. All inputs are plain data from the sample row.
/// Returns `NotApplicable` for non-CALLS categories.
///
/// This function is re-exported as `pub` from the crate root
/// because it is one of the 6 locked public entry points.
pub fn derive_blast_radius(
	category: UnresolvedEdgeCategory,
	basis_code: UnresolvedEdgeBasisCode,
	source_node_visibility: Option<&str>,
) -> BlastRadiusAssessment {
	if !category.is_calls_category() {
		return BlastRadiusAssessment {
			receiver_origin: ReceiverOrigin::NotApplicable,
			enclosing_scope_significance: EnclosingScopeSignificance::UnknownScope,
			blast_radius: BlastRadiusLevel::NotApplicable,
		};
	}

	let origin = derive_receiver_origin(basis_code);
	let scope = derive_enclosing_scope_significance(source_node_visibility);
	let radius = compute_radius(origin, scope);

	BlastRadiusAssessment {
		receiver_origin: origin,
		enclosing_scope_significance: scope,
		blast_radius: radius,
	}
}

/// Map basis_code to receiver origin.
///
/// Mirrors `deriveReceiverOrigin` from `blast-radius.ts:106`.
fn derive_receiver_origin(
	basis_code: UnresolvedEdgeBasisCode,
) -> ReceiverOrigin {
	match basis_code {
		UnresolvedEdgeBasisCode::ThisReceiverImpliesInternal => {
			ReceiverOrigin::ThisBound
		}
		UnresolvedEdgeBasisCode::ReceiverMatchesExternalImport
		| UnresolvedEdgeBasisCode::CalleeMatchesExternalImport => {
			ReceiverOrigin::ImportBoundExternal
		}
		UnresolvedEdgeBasisCode::ReceiverMatchesInternalImport
		| UnresolvedEdgeBasisCode::CalleeMatchesInternalImport
		| UnresolvedEdgeBasisCode::SpecifierMatchesProjectAlias => {
			ReceiverOrigin::ImportBoundInternal
		}
		UnresolvedEdgeBasisCode::ReceiverMatchesSameFileSymbol
		| UnresolvedEdgeBasisCode::CalleeMatchesSameFileSymbol => {
			ReceiverOrigin::SameFileDeclared
		}
		UnresolvedEdgeBasisCode::ReceiverMatchesRuntimeGlobal
		| UnresolvedEdgeBasisCode::CalleeMatchesRuntimeGlobal => {
			ReceiverOrigin::RuntimeGlobal
		}
		UnresolvedEdgeBasisCode::SpecifierMatchesRuntimeModule => {
			ReceiverOrigin::RuntimeModule
		}
		UnresolvedEdgeBasisCode::NoSupportingSignal => {
			ReceiverOrigin::LocalLike
		}
		// All other basis codes (package dep, relative import,
		// Rust heuristic, Express, etc.) fall to local_like.
		// Mirrors the TS `default: return "local_like"` case.
		_ => ReceiverOrigin::LocalLike,
	}
}

/// Map source node visibility to enclosing-scope significance.
///
/// Mirrors `deriveEnclosingScopeSignificance` from
/// `blast-radius.ts:132`.
fn derive_enclosing_scope_significance(
	visibility: Option<&str>,
) -> EnclosingScopeSignificance {
	match visibility {
		Some("export") | Some("public") => {
			EnclosingScopeSignificance::ExportedPublic
		}
		Some("private") | Some("protected") | Some("internal") | None => {
			EnclosingScopeSignificance::InternalPrivate
		}
		_ => EnclosingScopeSignificance::UnknownScope,
	}
}

/// Combine receiver origin + enclosing scope into a blast-radius
/// level.
///
/// Mirrors `computeRadius` from `blast-radius.ts:164`.
///
/// Principles:
///   - `local_like + internal_private` = LOW (function-local noise)
///   - `local_like + exported_public` = MEDIUM (local uncertainty
///     on a public path)
///   - `import_bound_*` = MEDIUM (cross-module reference) or LOW
///     (external — well-understood)
///   - `runtime_global / runtime_module` = LOW
///   - `same_file_declared` = LOW
///   - `this_bound + internal` = LOW, `this_bound + exported` = MEDIUM
///
/// Entrypoint-path detection (→ HIGH override) is deferred. No
/// edge currently reaches HIGH.
fn compute_radius(
	origin: ReceiverOrigin,
	scope: EnclosingScopeSignificance,
) -> BlastRadiusLevel {
	match origin {
		ReceiverOrigin::LocalLike | ReceiverOrigin::ThisBound => {
			if scope == EnclosingScopeSignificance::ExportedPublic {
				BlastRadiusLevel::Medium
			} else {
				BlastRadiusLevel::Low
			}
		}
		ReceiverOrigin::ImportBoundInternal => BlastRadiusLevel::Medium,
		ReceiverOrigin::ImportBoundExternal
		| ReceiverOrigin::SameFileDeclared
		| ReceiverOrigin::RuntimeGlobal
		| ReceiverOrigin::RuntimeModule => BlastRadiusLevel::Low,
		ReceiverOrigin::NotApplicable => BlastRadiusLevel::NotApplicable,
	}
}

#[cfg(test)]
mod tests {
	use super::*;

	// ── Receiver origin mapping ──────────────────────────────

	fn calls_category() -> UnresolvedEdgeCategory {
		UnresolvedEdgeCategory::CallsObjMethodNeedsTypeInfo
	}

	#[test]
	fn this_receiver_implies_internal_maps_to_this_bound() {
		let r = derive_blast_radius(
			calls_category(),
			UnresolvedEdgeBasisCode::ThisReceiverImpliesInternal,
			None,
		);
		assert_eq!(r.receiver_origin, ReceiverOrigin::ThisBound);
	}

	#[test]
	fn callee_matches_external_import_maps_to_import_bound_external() {
		let r = derive_blast_radius(
			calls_category(),
			UnresolvedEdgeBasisCode::CalleeMatchesExternalImport,
			None,
		);
		assert_eq!(r.receiver_origin, ReceiverOrigin::ImportBoundExternal);
	}

	#[test]
	fn receiver_matches_internal_import_maps_to_import_bound_internal() {
		let r = derive_blast_radius(
			calls_category(),
			UnresolvedEdgeBasisCode::ReceiverMatchesInternalImport,
			None,
		);
		assert_eq!(r.receiver_origin, ReceiverOrigin::ImportBoundInternal);
	}

	#[test]
	fn specifier_matches_project_alias_maps_to_import_bound_internal() {
		let r = derive_blast_radius(
			calls_category(),
			UnresolvedEdgeBasisCode::SpecifierMatchesProjectAlias,
			None,
		);
		assert_eq!(r.receiver_origin, ReceiverOrigin::ImportBoundInternal);
	}

	#[test]
	fn callee_matches_same_file_symbol_maps_to_same_file_declared() {
		let r = derive_blast_radius(
			calls_category(),
			UnresolvedEdgeBasisCode::CalleeMatchesSameFileSymbol,
			None,
		);
		assert_eq!(r.receiver_origin, ReceiverOrigin::SameFileDeclared);
	}

	#[test]
	fn callee_matches_runtime_global_maps_to_runtime_global() {
		let r = derive_blast_radius(
			calls_category(),
			UnresolvedEdgeBasisCode::CalleeMatchesRuntimeGlobal,
			None,
		);
		assert_eq!(r.receiver_origin, ReceiverOrigin::RuntimeGlobal);
	}

	#[test]
	fn specifier_matches_runtime_module_maps_to_runtime_module() {
		let r = derive_blast_radius(
			calls_category(),
			UnresolvedEdgeBasisCode::SpecifierMatchesRuntimeModule,
			None,
		);
		assert_eq!(r.receiver_origin, ReceiverOrigin::RuntimeModule);
	}

	#[test]
	fn no_supporting_signal_maps_to_local_like() {
		let r = derive_blast_radius(
			calls_category(),
			UnresolvedEdgeBasisCode::NoSupportingSignal,
			None,
		);
		assert_eq!(r.receiver_origin, ReceiverOrigin::LocalLike);
	}

	// ── Enclosing scope significance ─────────────────────────

	#[test]
	fn export_visibility_is_exported_public() {
		let r = derive_blast_radius(
			calls_category(),
			UnresolvedEdgeBasisCode::NoSupportingSignal,
			Some("export"),
		);
		assert_eq!(
			r.enclosing_scope_significance,
			EnclosingScopeSignificance::ExportedPublic
		);
	}

	#[test]
	fn private_visibility_is_internal_private() {
		let r = derive_blast_radius(
			calls_category(),
			UnresolvedEdgeBasisCode::NoSupportingSignal,
			Some("private"),
		);
		assert_eq!(
			r.enclosing_scope_significance,
			EnclosingScopeSignificance::InternalPrivate
		);
	}

	#[test]
	fn null_visibility_is_internal_private() {
		let r = derive_blast_radius(
			calls_category(),
			UnresolvedEdgeBasisCode::NoSupportingSignal,
			None,
		);
		assert_eq!(
			r.enclosing_scope_significance,
			EnclosingScopeSignificance::InternalPrivate
		);
	}

	// ── Radius computation ───────────────────────────────────

	#[test]
	fn local_like_internal_private_is_low() {
		let r = derive_blast_radius(
			calls_category(),
			UnresolvedEdgeBasisCode::NoSupportingSignal,
			Some("private"),
		);
		assert_eq!(r.blast_radius, BlastRadiusLevel::Low);
	}

	#[test]
	fn local_like_exported_public_is_medium() {
		let r = derive_blast_radius(
			calls_category(),
			UnresolvedEdgeBasisCode::NoSupportingSignal,
			Some("export"),
		);
		assert_eq!(r.blast_radius, BlastRadiusLevel::Medium);
	}

	#[test]
	fn this_bound_internal_private_is_low() {
		let r = derive_blast_radius(
			calls_category(),
			UnresolvedEdgeBasisCode::ThisReceiverImpliesInternal,
			Some("private"),
		);
		assert_eq!(r.blast_radius, BlastRadiusLevel::Low);
	}

	#[test]
	fn this_bound_exported_public_is_medium() {
		let r = derive_blast_radius(
			calls_category(),
			UnresolvedEdgeBasisCode::ThisReceiverImpliesInternal,
			Some("export"),
		);
		assert_eq!(r.blast_radius, BlastRadiusLevel::Medium);
	}

	#[test]
	fn import_bound_internal_is_medium_regardless_of_scope() {
		let r = derive_blast_radius(
			calls_category(),
			UnresolvedEdgeBasisCode::ReceiverMatchesInternalImport,
			Some("private"),
		);
		assert_eq!(r.blast_radius, BlastRadiusLevel::Medium);
	}

	#[test]
	fn import_bound_external_is_low() {
		let r = derive_blast_radius(
			calls_category(),
			UnresolvedEdgeBasisCode::CalleeMatchesExternalImport,
			Some("export"),
		);
		assert_eq!(r.blast_radius, BlastRadiusLevel::Low);
	}

	#[test]
	fn same_file_declared_is_low() {
		let r = derive_blast_radius(
			calls_category(),
			UnresolvedEdgeBasisCode::CalleeMatchesSameFileSymbol,
			Some("export"),
		);
		assert_eq!(r.blast_radius, BlastRadiusLevel::Low);
	}

	#[test]
	fn runtime_global_is_low() {
		let r = derive_blast_radius(
			calls_category(),
			UnresolvedEdgeBasisCode::CalleeMatchesRuntimeGlobal,
			Some("export"),
		);
		assert_eq!(r.blast_radius, BlastRadiusLevel::Low);
	}

	// ── Non-CALLS categories ─────────────────────────────────

	#[test]
	fn imports_category_is_not_applicable() {
		let r = derive_blast_radius(
			UnresolvedEdgeCategory::ImportsFileNotFound,
			UnresolvedEdgeBasisCode::NoSupportingSignal,
			Some("export"),
		);
		assert_eq!(r.blast_radius, BlastRadiusLevel::NotApplicable);
		assert_eq!(r.receiver_origin, ReceiverOrigin::NotApplicable);
	}

	#[test]
	fn instantiates_category_is_not_applicable() {
		let r = derive_blast_radius(
			UnresolvedEdgeCategory::InstantiatesClassNotFound,
			UnresolvedEdgeBasisCode::NoSupportingSignal,
			None,
		);
		assert_eq!(r.blast_radius, BlastRadiusLevel::NotApplicable);
	}

	#[test]
	fn other_category_is_not_applicable() {
		let r = derive_blast_radius(
			UnresolvedEdgeCategory::Other,
			UnresolvedEdgeBasisCode::NoSupportingSignal,
			None,
		);
		assert_eq!(r.blast_radius, BlastRadiusLevel::NotApplicable);
	}
}
